pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,539 @@
+};
+use cfg_types::domain_address::{Domain, DomainAddress};
+use frame_support::{
+	dispatch::DispatchResult,
+	ensure,
+	pallet_prelude::{Decode, Encode, Get, TypeInfo},
+	weights::Weight,
+};
+use parity_scale_codec::MaxEncodedLen;
+use sp_arithmetic::traits::{EnsureAddAssign, EnsureSub, SaturatedConversion};
+use sp_runtime::DispatchError;
+
+use crate::{
+	message::GatewayMessage, Config, Error, Pallet, PendingInboundEntries, Routers, SessionIdStore,
+};
+
+/// Type that holds the information needed for inbound message entries.
+#[derive(Debug, Encode, Decode, Clone, Eq, MaxEncodedLen, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct MessageEntry<T: Config> {
+	/// The session ID for this entry.
+	pub session_id: T::SessionId,
+
+	/// The sender of the inbound message.
+	pub domain_address: DomainAddress,
+
+	/// The LP message.
+	pub message: T::Message,
+
+	/// The expected proof count for processing one or more of the provided
+	/// message.
+	///
+	/// NOTE - this gets increased by the `expected_proof_count` for a set of
+	/// routers (see `get_expected_proof_count`) every time a new identical
+	/// message is submitted.
+	pub expected_proof_count: u32,
+}
+
+/// Type that holds the information needed for inbound proof entries.
+#[derive(Debug, Encode, Decode, Clone, Eq, MaxEncodedLen, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct ProofEntry<T: Config> {
+	/// The session ID for this entry.
+	pub session_id: T::SessionId,
+
+	/// The number of proofs received for a particular message.
+	///
+	/// NOTE - this gets increased by 1 every time a new identical message is
+	/// submitted.
+	pub current_count: u32,
+}
+
+impl<T: Config> ProofEntry<T> {
+	/// Returns `true` if all the following conditions are true:
+	/// - the session IDs match
+	/// - the `current_count` is greater than 0
+	pub fn has_valid_vote_for_session(&self, session_id: T::SessionId) -> bool {
+		self.session_id == session_id && self.current_count > 0
+	}
+}
+
+/// Type used when storing inbound message information.
+#[derive(Debug, Encode, Decode, Clone, Eq, MaxEncodedLen, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub enum InboundEntry<T: Config> {
+	Message(MessageEntry<T>),
+	Proof(ProofEntry<T>),
+}
+
+impl<T: Config> From<(&InboundProcessingInfo<T>, T::Message)> for InboundEntry<T> {
+	fn from((inbound_processing_info, message): (&InboundProcessingInfo<T>, T::Message)) -> Self {
+		match message.get_message_proof() {
+			None => InboundEntry::Message(MessageEntry {
+				session_id: inbound_processing_info.current_session_id.clone(),
+				domain_address: inbound_processing_info.domain_address.clone(),
+				message,
+				expected_proof_count: inbound_processing_info.expected_proof_count_per_message,
+			}),
+			Some(_) => InboundEntry::Proof(ProofEntry {
+				session_id: inbound_processing_info.current_session_id,
+				current_count: 1,
+			}),
+		}
+	}
+}
```

Comment:
> As a note, From/TryFrom are very cool but they're a double-edged sword, because you lose traceability. They are usually used only for very isometric types.

Definitely agree. Feels like having a dedicated fuction such as `into_inbound_entry` is clearer. Though this is definitely a nit! Since we have more `From` impls below, no need to change this for now IMO.

Reply:
> Since we have more From impls below

I think the below ones makes sense because they are straightforward, just wrappers and things like that

Reply:
I'll admit that I used and abused from/try_from :D. Dropped those and now there's a specific `create` fn in the `InboundEntry` impl.